File: python/examples/async-limit-concurrency_test.py
"""async-limit-concurrency: asyncio.Semaphore の Contract を検証する。"""

import asyncio

import pytest


def test_limits_concurrent_tasks() -> None:
    """async with sem で同時実行数が value 以下に抑えられ、抜けるときに解放される。"""
    sem = asyncio.Semaphore(2)
    running = 0
    peak = 0

    async def work() -> None:
        nonlocal running, peak
        async with sem:
            running += 1
            peak = max(peak, running)
            await asyncio.sleep(0.01)
            running -= 1

    async def main() -> None:
        await asyncio.gather(*(work() for _ in range(6)))

    asyncio.run(main())
    assert peak == 2
    assert sem.locked() is False


def test_releases_even_when_block_raises() -> None:
    """ブロックの中で例外が出ても解放される。"""
    sem = asyncio.Semaphore(1)

    async def main() -> None:
        with pytest.raises(ValueError, match="inside"):
            async with sem:
                raise ValueError("inside")
        assert sem.locked() is False
        async with sem:
            assert sem.locked() is True

    asyncio.run(main())


def test_acquire_returns_immediately_when_available() -> None:
    """空きがあれば acquire は即座に戻り、空きが無ければ release まで待つ。"""
    order: list[str] = []

    async def main() -> None:
        sem = asyncio.Semaphore(1)
        assert await sem.acquire() is True
        assert sem.locked() is True

        async def waiter() -> None:
            order.append("waiting")
            await sem.acquire()
            order.append("acquired")

        task = asyncio.create_task(waiter())
        await asyncio.sleep(0.01)
        assert order == ["waiting"]
        sem.release()
        await task
        assert order == ["waiting", "acquired"]

    asyncio.run(main())


def test_all_waiters_complete_after_release() -> None:
    """release() すると待機中の全タスクが（順序は問わず）完了する。"""
    order: list[int] = []

    async def main() -> None:
        sem = asyncio.Semaphore(1)
        await sem.acquire()

        async def waiter(n: int) -> None:
            async with sem:
                order.append(n)

        tasks = [asyncio.create_task(waiter(n)) for n in (3, 1, 2)]
        await asyncio.sleep(0)
        sem.release()
        await asyncio.gather(*tasks)

    asyncio.run(main())
    # 再開順は仕様で保証されない（CPython では acquire 順）。全員が完了したことだけを検証する
    assert sorted(order) == [1, 2, 3]


def test_extra_release_raises_the_limit() -> None:
    """release は初期値を超えても例外にならず、その分だけ同時実行数が増える。"""
    sem = asyncio.Semaphore(1)
    sem.release()
    running = 0
    peak = 0

    async def work() -> None:
        nonlocal running, peak
        async with sem:
            running += 1
            peak = max(peak, running)
            await asyncio.sleep(0.01)
            running -= 1

    async def main() -> None:
        await asyncio.gather(work(), work(), work())

    asyncio.run(main())
    assert peak == 2


def test_locked_reflects_availability() -> None:
    """locked() は空きが無いときだけ True。"""
    sem = asyncio.Semaphore(0)
    assert sem.locked() is True
    sem.release()
    assert sem.locked() is False


def test_cancelled_waiter_does_not_take_permit() -> None:
    """待機中の acquire がキャンセルされると CancelledError になり、permit は取得しない。"""

    async def main() -> None:
        sem = asyncio.Semaphore(1)
        await sem.acquire()
        task = asyncio.create_task(sem.acquire())
        await asyncio.sleep(0)
        task.cancel()
        with pytest.raises(asyncio.CancelledError):
            await task
        sem.release()
        assert sem.locked() is False

    asyncio.run(main())


def test_negative_initial_value_raises() -> None:
    """初期値に負数を渡すと ValueError。"""
    with pytest.raises(ValueError):
        asyncio.Semaphore(-1)


def test_bounded_semaphore_rejects_extra_release() -> None:
    """Alternatives: BoundedSemaphore は余分な release を ValueError で検出する。"""
    sem = asyncio.BoundedSemaphore(1)
    with pytest.raises(ValueError):
        sem.release()

File: typescript/examples/async-limit-concurrency.test.ts
// カード async-limit-concurrency の Contract を検証するテスト
import { Semaphore } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

/** マイクロタスクを消化して待機中の Promise を進める */
const flush = () => new Promise<void>((resolve) => queueMicrotask(resolve));

describe('async-limit-concurrency: Semaphore', () => {
  it('空きがあれば即 resolve し、available が減る', async () => {
    const sem = new Semaphore(2);
    expect(sem.capacity).toBe(2);
    expect(sem.available).toBe(2);
    await sem.acquire();
    expect(sem.available).toBe(1);
  });

  it('空きが無ければ release されるまで待つ', async () => {
    const sem = new Semaphore(1);
    await sem.acquire();
    let acquired = false;
    const waiting = sem.acquire().then(() => {
      acquired = true;
    });
    await flush();
    expect(acquired).toBe(false);
    sem.release();
    await waiting;
    expect(acquired).toBe(true);
    expect(sem.available).toBe(0); // permit は待機者へ直接渡り、available は変わらない
  });

  it('待機は呼び出し順（FIFO）に解放される', async () => {
    const sem = new Semaphore(1);
    await sem.acquire();
    const order: string[] = [];
    const a = sem.acquire().then(() => order.push('a'));
    const b = sem.acquire().then(() => order.push('b'));
    sem.release();
    await a;
    expect(order).toEqual(['a']);
    sem.release();
    await b;
    expect(order).toEqual(['a', 'b']);
  });

  it('release は待機者がいなければ available を増やし、capacity を超えない', async () => {
    const sem = new Semaphore(2);
    await sem.acquire();
    sem.release();
    expect(sem.available).toBe(2);
    sem.release(); // 余分な release は無視される
    expect(sem.available).toBe(2);
  });

  it('同時実行数が capacity を超えない', async () => {
    const sem = new Semaphore(2);
    let running = 0;
    let peak = 0;
    const task = async () => {
      await sem.acquire();
      try {
        running++;
        peak = Math.max(peak, running);
        await flush();
        running--;
      } finally {
        sem.release();
      }
    };
    await Promise.all([task(), task(), task(), task(), task()]);
    expect(peak).toBe(2);
    expect(sem.available).toBe(2);
  });

  it('release し忘れると以後の acquire は resolve しない', async () => {
    const sem = new Semaphore(1);
    await sem.acquire();
    let acquired = false;
    void sem.acquire().then(() => {
      acquired = true;
    });
    await flush();
    await flush();
    expect(acquired).toBe(false);
  });
});
